File: src/test/java/ukalus/path/astar/AStarTest.java
/*
 * Created on Feb 15, 2005
 *
 */
package ukalus.path.astar;

import ukalus.level.Level;
import ukalus.math.Vector2D;

import java.util.Iterator;

import junit.framework.TestCase;

/**
 * @author gremopm
 * 
 */
public class AStarTest extends TestCase {

  public void testFindPathNoObstacles() {
    Level map = new Level(new Object[][]{
        {Boolean.TRUE, Boolean.TRUE, Boolean.TRUE},
        {Boolean.TRUE, Boolean.TRUE, Boolean.TRUE},
        {Boolean.TRUE, Boolean.TRUE, Boolean.TRUE}});
    AStar finder = new AStar();
    Cost cost = new FixedCost(1);
    Node2D start = new Node2D(map, Vector2D.get(0, 0), null);
    Node2D stop = new Node2D(map, Vector2D.get(2, 2), null);
    Heuristic heuristic = new ManhattenHeuristic(Vector2D.get(0, 0),
      Vector2D.get(2, 2));
    Iterator<Node> path = finder.solve(heuristic, cost, start, stop);
    assertTrue(path.hasNext());

    assertEquals(Vector2D.get(0, 0), ((Node2D) path.next()).getLocation());
    assertEquals(Vector2D.get(1, 0), ((Node2D) path.next()).getLocation());
    assertEquals(Vector2D.get(1, 1), ((Node2D) path.next()).getLocation());
    assertEquals(Vector2D.get(2, 1), ((Node2D) path.next()).getLocation());
    assertEquals(Vector2D.get(2, 2), ((Node2D) path.next()).getLocation());

  }

  public void testFindPathChoke() {
    AStar finder = new AStar();
    Cost cost = new FixedCost(1);
    Level map = new Level(new Object[][]{
        {Boolean.TRUE, null, Boolean.TRUE},
        {Boolean.TRUE, Boolean.TRUE, Boolean.TRUE},
        {Boolean.TRUE, null, Boolean.TRUE}});
    Node2D start = new Node2D(map, Vector2D.get(0, 0), null);
    Node2D stop = new Node2D(map, Vector2D.get(2, 2), null);
    Heuristic heuristic = new ManhattenHeuristic(Vector2D.get(0, 0),
      Vector2D.get(2, 2));
    Iterator<Node> path = finder.solve(heuristic, cost, start, stop);
    assertTrue(path.hasNext());

    assertEquals(Vector2D.get(0, 0), ((Node2D) path.next()).getLocation());
    assertEquals(Vector2D.get(1, 0), ((Node2D) path.next()).getLocation());
    assertEquals(Vector2D.get(1, 1), ((Node2D) path.next()).getLocation());
    assertEquals(Vector2D.get(1, 2), ((Node2D) path.next()).getLocation());
    assertEquals(Vector2D.get(2, 2), ((Node2D) path.next()).getLocation());
  }
}

File: src/test/kotlin/ukalus/path/astar/AStarTest.kt
/*
 * Created on Feb 15, 2005
 *
 */
package ukalus.path.astar

import junit.framework.TestCase
import ukalus.level.Level
import ukalus.math.Vector2D
import java.lang.Boolean.TRUE

/**
 * @author gremopm
 */
class AStarTest : TestCase() {

    fun testFindPathNoObstacles() {
        val heuristic = manhattanHeuristic(Vector2D(0, 0), Vector2D(2, 2))
        val cost = fixedCost(1)
        val map = Level(arrayOf(arrayOf(TRUE, TRUE, TRUE), arrayOf(TRUE, TRUE, TRUE), arrayOf(TRUE, TRUE, TRUE)))
        val start = Node2D(map, Vector2D(0, 0), null)
        val stop = Node2D(map, Vector2D(2, 2), null)
        val path = AStar().solve(heuristic, cost, start, stop).map { (it as Node2D).location }

        assertEquals(listOf(Vector2D(0, 0), Vector2D(1, 0), Vector2D(1, 1), Vector2D(1, 2), Vector2D(2, 2)), path)
    }

    fun testFindPathChoke() {
        val heuristic = manhattanHeuristic(Vector2D(0, 0), Vector2D(2, 2))
        val cost = fixedCost(1)
        val map = Level(arrayOf(arrayOf(TRUE, null, TRUE), arrayOf(TRUE, TRUE, TRUE), arrayOf(TRUE, null, TRUE)))
        val start = Node2D(map, Vector2D(0, 0), null)
        val stop = Node2D(map, Vector2D(2, 2), null)
        val path = AStar().solve(heuristic, cost, start, stop).map { (it as Node2D).location }

        assertEquals(listOf(Vector2D(0, 0), Vector2D(1, 0), Vector2D(1, 1), Vector2D(1, 2), Vector2D(2, 2)), path)
    }
}